ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -733,53 +760,89 @@ private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcInd
     long totalCapacity = tgtLog.getCapacityInBytes();
     long writtenLastTime = 0;
     try (FileChannel fileChannel = Utils.openChannel(logSegmentToCopy.getView().getFirst(), false)) {
-      for (IndexEntry srcIndexEntry : srcIndexEntries) {
-        IndexValue srcValue = srcIndexEntry.getValue();
-        long usedCapacity = tgtIndex.getLogUsedCapacity();
-        if (isActive && (tgtLog.getCapacityInBytes() - usedCapacity >= srcValue.getSize())) {
-          fileChannel.position(srcValue.getOffset().getOffset());
-          Offset endOffsetOfLastMessage = tgtLog.getEndOffset();
-          // call into diskIOScheduler to make sure we can proceed (assuming it won't be 0).
-          diskIOScheduler.getSlice(DiskManager.CLEANUP_OPS_JOB_NAME, DiskManager.CLEANUP_OPS_JOB_NAME, writtenLastTime);
-          tgtLog.appendFrom(fileChannel, srcValue.getSize());
-          FileSpan fileSpan = tgtLog.getFileSpanForMessage(endOffsetOfLastMessage, srcValue.getSize());
-          if (srcValue.isFlagSet(IndexValue.Flags.Delete_Index)) {
-            IndexValue putValue = tgtIndex.findKey(srcIndexEntry.getKey());
-            if (putValue != null) {
-              tgtIndex.markAsDeleted(srcIndexEntry.getKey(), fileSpan, srcValue.getOperationTimeInMs());
+      int startIndex = 0;
+      while (startIndex < srcIndexEntries.size()) {
+        // try to do a bundle of read to reduce disk IO
+        ByteBuffer bufferToUse;
+        long startOffset = srcIndexEntries.get(startIndex).getValue().getOffset().getOffset();
+        int endIndex;
+        if (bundleReadBuffer == null
+            || srcIndexEntries.get(startIndex).getValue().getSize() > bundleReadBuffer.capacity()) {
+          // can't use the preAllocated buffer
+          endIndex = startIndex;
```

Comment:
maybe a metric (and a trace log message with the id and the size) here will be useful